### PR TITLE
Support a query with multiple statments (unofficially)

### PR DESCRIPTION
**Pull request by sonots**
Fix https://github.com/wvanbergen/vertica/issues/63

### CODE REVIEW

File: lib/vertica/query.rb [hunk 1]
```diff
@@ -18,7 +18,8 @@ class Vertica::Query
   #   to receive data for a `COPY ... FROM STDIN` statement.
   def initialize(connection, sql, row_handler: nil, copy_handler: nil)
     @connection, @sql = connection, sql
-    @buffer = row_handler.nil? && copy_handler.nil? ? [] : nil
+    @reset_buffer_handler = lambda { row_handler.nil? && copy_handler.nil? ? [] : nil }
```
Comment (wvanbergen) [inline lib/vertica/query.rb]: Can you change this into a private method rather than a lambda?

File: test/functional/functional_query_test.rb [hunk 2]
```diff
@@ -81,6 +81,15 @@ def test_select_query_with_multibyte_column_names_results
...
   end
 
+  # Query with multiple statements is not officailly supported
+  # Results of only the last statement are obtained
+  def test_query_with_multiple_statements
+    r = @connection.query("SET SEARCH_PATH TO default; SELECT 1 as one")
+    assert_equal 1, r.size
+    assert_equal 1, r.row_description.length
+    assert_equal 'one', r.row_description[0].name
+  end
```
Comment (wvanbergen) [inline test/functional/functional_query_test.rb]: Can you also add a test that documents the behaviour if you use two queries that do return a resultset, e.g. `SELECT 1; SELECT 2`

File: lib/vertica/query.rb (hunk 1 above)
Comment (sonots) [inline lib/vertica/query.rb]: To do it, I have to store the original row_handler, and copy_handler although @raw_handler would be a lambda created in the constructor. Do you prefer it?

Comment (sonots) [inline lib/vertica/query.rb]: Anyway, changed so.

Comment (wvanbergen) [inline lib/vertica/query.rb]: Maybe something like this?

``` ruby
enable_buffering = row_handler.nil? && copy_handler.nil?
@buffer = [] if enable_buffering

# to reset
@buffer.clear if defined(@buffer)
```

Comment (sonots) [inline lib/vertica/query.rb]: Fixed like that :-)

File: lib/vertica/query.rb [hunk 3]
```diff
@@ -83,7 +84,8 @@ def handle_data_row(message)
   def handle_command_complete(message)
     if buffer_rows?
       @result = Vertica::Result.new(row_description: @row_description, rows: @buffer, tag: message.tag)
-      @row_description, @buffer = nil, nil
+      @row_description = nil
+      @buffer.clear if @enable_buffering
```
Comment (wvanbergen) [inline lib/vertica/query.rb]: Should we also reset `@error` here?

File: test/functional/functional_query_test.rb (hunk 2 above)
Comment (sonots) [inline test/functional/functional_query_test.rb]: I changed the test query to `"SET SEARCH_PATH TO default; SELECT 1 as one; SELECT 2 as two"`.

File: lib/vertica/query.rb [hunk 4]
```diff
@@ -18,7 +18,8 @@ class Vertica::Query
   #   to receive data for a `COPY ... FROM STDIN` statement.
   def initialize(connection, sql, row_handler: nil, copy_handler: nil)
     @connection, @sql = connection, sql
-    @buffer = row_handler.nil? && copy_handler.nil? ? [] : nil
+    @enable_buffering = row_handler.nil? && copy_handler.nil?
```
Comment (wvanbergen) [inline lib/vertica/query.rb]: I notice we have `buffer_rows?`already. Can we make use of that, or change the implementation to  simply return `@enable_buffering`?

Comment (sonots) [inline lib/vertica/query.rb]: Agree, fixed to use `buffer_rows?`

File: lib/vertica/query.rb (hunk 3 above)
Comment (sonots) [inline lib/vertica/query.rb]: I want to get an error if a former statement gets errored, and currently it is working so. I added tests to show the spec.

File: lib/vertica/query.rb [hunk 5]
```diff
@@ -83,7 +83,8 @@ def handle_data_row(message)
   def handle_command_complete(message)
     if buffer_rows?
       @result = Vertica::Result.new(row_description: @row_description, rows: @buffer, tag: message.tag)
-      @row_description, @buffer = nil, nil
+      @row_description = nil
+      @buffer.clear if buffer_rows?
```
Comment (wvanbergen) [inline lib/vertica/query.rb]: I just noticed that we're already in an `if buffer_rows?` block (see line 83). I can we can just always just clear it here and get rid of the if suffix.

Comment (sonots) [inline lib/vertica/query.rb]: right

File: test/functional/functional_query_test.rb [hunk 6]
```diff
@@ -81,6 +81,20 @@ def test_select_query_with_multibyte_column_names_results
...
+    r = @connection.query("SET SEARCH_PATH TO default; SELECT 1 as one; SELECT 2 as two")
+    assert_equal 1, r.size
+    assert_equal 1, r.row_description.length
+    assert_equal 'two', r.row_description[0].name
+
+    # Error of the former statement is raised
+    assert_raises(Vertica::Error::MissingColumn) do
+      @connection.query("SELECT missing FROM test_ruby_vertica_table; SELECT 1 FROM missing")
+    end
+  end
```
Comment (wvanbergen) [inline test/functional/functional_query_test.rb]: Final question, mostly for my curiosity: 
- Does it still run the second query if thew first one fails?
- If so, does it send data to us? Is the connection in a usable state after this?

**Comment by wvanbergen** [comment]
Sorry for all the back and forth, but I think this is looking pretty good? Thanks for working on this ❤️

File: test/functional/functional_query_test.rb (hunk 6 above)
Comment (sonots) [inline test/functional/functional_query_test.rb]: No, messages become as

```
#<Vertica::Protocol::ErrorResponse:0x007f143ff48288>
#<Vertica::Protocol::ReadyForQuery:0x007f143ff43670>
```

Comment (wvanbergen) [inline test/functional/functional_query_test.rb]: :+1: awesome

Comment (sonots) [inline test/functional/functional_query_test.rb]: More info, `select 1; select missing; select 1` becomes as

```
#<Vertica::Protocol::RowDescription:0x007f143fef5880>
#<Vertica::Protocol::DataRow:0x007f143fef5330>
#<Vertica::Protocol::CommandComplete:0x007f143fef4e30>
#<Vertica::Protocol::ErrorResponse:0x007f143fef4b10>
#<Vertica::Protocol::ReadyForQuery:0x007f143fef3da0>
```

**Comment by wvanbergen** [comment]
Just waiting for CI now...

File: lib/vertica/query.rb (hunk 5 above)
Comment (sonots) [inline lib/vertica/query.rb]: `clear` was not good because the result `@buffer` is cleared here. We should set a new array as `@buffer = []`

**Comment by sonots** [comment]
CI passed

**Pull request merged**